### ISSUE TITLE
GD-168: Incorrect assertion messages

### ISSUE THREAD

**Issue by esainane** [MikeSchulze/gdUnit4Net#168]
### The used GdUnit4 version

4.4.4 (Pre Release/Master branch)

### The used Godot version

v4.3.1.rc.mono.custom_build [ff9bc0422]

### Operating System

`6.11.5+bpo-amd64 #1 SMP PREEMPT_DYNAMIC Debian GNU/Linux trixie/sid 6.11.5-1~bpo12+1 (2024-11-11) x86_64`

### Describe the bug

When an assertion fails, some error messages place the `actual` value after the `expected:` marker, and the `expected` value after the `but was` marker. The `expected` value should come after `Expected:`.

For example, this current code prints `Expecting: <expected> not equal to <actual>`, as expected:

```gd
static func error_not_equal(current :Variant, expected :Variant) -> String:
	return "%s\n %s\n not equal to\n %s" % [_error("Expecting:"), _colored_value(expected), _colored_value(current)]
```

However, this current code prints messages with the `actual` and `expected` values reversed, in the form `Unexpected type comparison: Expecting type <actual> but is <expected>`:

```gd
static func error_is_wrong_type(current_type :Variant.Type, expected_type :Variant.Type) -> String:
	return "%s\n Expecting type %s but is %s" % [
		_error("Unexpected type comparison:"),
		_colored_value(GdObjects.type_as_string(current_type)),
		_colored_value(GdObjects.type_as_string(expected_type))]
```

Many other functions also have this problem. For example, all of the array assert functions have the `actual` and `expected` parameters swapped in their error message.

### Steps to Reproduce

Create a small test case with a failing array assertion:

```cs
namespace GDUnit4AssertionTestExample;

using Godot;
using GdUnit4;

using static GdUnit4.Assertions;

[TestSuite]
public class ExampleTest
{
    [TestCase]
    public void FailMe() => AssertArray(new[] { "foo" }).ContainsExactly(new[] { "bar" });
}
```

This produces the following error:

![image](https://github.com/user-attachments/assets/099ff4b0-afab-43c1-af79-62f32bcba61f)

### Linked Issues
[assert_string](https://github.com/MikeSchulze/gdUnit4/issues/611)

**Comment by MikeSchulze** [comment]
@esainane can you confirm this?

**Comment by esainane** [comment]
First, version check. When running in VS Code:

```
CheckGdUnit4ApiVersion gdUnit4Api, Version=4.4.0.0
```

I cannot seem to find a test log in the godot editor, but the panel says it is `gdUnit4 4.4.4`.

---

On the **order** topic, you are right about the array assertion example. I made a mistake when porting the test to GDScript. The order is:
```
line 10: Expecting contains exactly elements:
 '["actual"]'
 do contains (in same order)
 '["expected"]'
but some elements where not expected:
 '["actual"]'
and could not find elements:
 '["expected"]'
``` 
in both CS and GDScript. I've attached a fixed test project here.

The overall message still does not make sense to me, as it seems to be listing "actual" as an element both expected (in the first printed value) and not expected (in the third printed value)? I think I'm still missing something here?


The mismatched type example has a clearer message, and the order does seem wrong here:

```gd
func test_assert_vector_typefail():
    var expected = Vector2(1, 2)
    var actual = false
    assert_vector(actual).is_equal(expected)
```
which produces:
![Screenshot from 2024-12-07 15-11-08](https://github.com/user-attachments/assets/7553f4e6-4d93-441b-91f6-603d361d03d0)
I would expect something like `"Expecting type 'Vector2' but was 'bool'"`?

---

On the **spacing** topic, this only shows up in a GDScript test, and does not seem to be related to any behaviour in VS Code, as I previously incorrectly assumed.

When running a C# test in the godot editor, the message is consistently aligned.

![Screenshot from 2024-12-07 15-10-55](https://github.com/user-attachments/assets/ae65b21c-66c5-49d4-89a9-4c20384a65a9)

When running a GDScript test in the godot editor, the second text line (`"do contains ..."`) has a leading space, but the others do not:

![Screenshot from 2024-12-07 15-10-47](https://github.com/user-attachments/assets/8f1bad90-6054-4702-979c-a939f14c0e1b)

[gdunit4-assert-test-example.zip](https://github.com/user-attachments/files/18048917/gdunit4-assert-test-example.zip)

**Comment by MikeSchulze** [comment]
Ok I'm on your side, there are some failure messages that not nice to read.
Before we start to refactor the failure messages, we need to define the pattern and build a list of them for each type.
Then we can refactor it by build a task for each assert and describe the pattern and expected messages.

Thanks for looking into this topic

**Comment by esainane** [comment]
Ah, my apologies if this came across as accusing or argumentative! I'm really grateful for gdUnit4, and wanted to help give back.

Is there a way I could help review testing assertion report formats? I'm most familiar with catch2, though I've also used jest, pytest, python-unittest, gtest, junit, and probably more which I'm not immediately remembering right now

**Comment by MikeSchulze** [comment]
All good ;), I am grateful for any help to improve the API.

The bug reports are heavily based on JUnit and should stay in that pattern.

If you really want to get involved here, I ask you to create new tickets for each assert type separately for GDScript and CSharp API. 
- Each ticket has a list of assert functions with an indication of consistent or should be revised. (checkmark)
- Add for each function a details section
  - function signature
  - test example
  - report output + screenshot
  - remarks (describe why the report is misleading and how it should look like)

## The pattern is

```
<details>
<summary>func_name</summary>

<function_signature>

<example_test>
  
### Failure Report
<report_text>
<report_screenshot>


### Remark
<your_remark>
</details>
```

An example issue for GDScript 'assert_string': (not fullfilled)
https://github.com/MikeSchulze/gdUnit4/issues/611

**Comment by esainane** [comment]
Hi, yes! Very sorry about the delay, I've been travelling and offline a lot lately. I should be able to take a look at this again tonight.

**Comment by MikeSchulze** [comment]
Please note, the master is actual under heavy development to introduce the new test engine.

**Comment by esainane** [comment]
If it would be inconvenient or distracting right now, I could also wait for a less busy time?

**Comment by MikeSchulze** [comment]
Yes 👍please wait until I get the first stable version of the new test engine running.

**Comment by esainane** [comment]
Will do!

(And as always, thank you for all of your hard work on gdUnit4!)
